library/src/blas1/rocblas_asum_batched.hpp

```diff
@@ -0,0 +1,56 @@
+/* ************************************************************************
+ * Copyright 2016-2019 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+#include "fetch_template.h"
+#include "handle.h"
+#include "reduction_batched.h"
+#include "rocblas.h"
+
+template <class To>
+struct rocblas_fetch_asum_batched
+{
+    template <typename Ti>
+    __forceinline__ __device__ To operator()(Ti x, ptrdiff_t)
+    {
+        return {fetch_asum(x)};
+    }
+};
+
+template <rocblas_int NB, typename To>
+size_t rocblas_asum_batched_template_workspace_size(rocblas_int n,
+                                                    rocblas_int batch_count,
+                                                    To*         output_type)
+{
+    return rocblas_reduction_batched_kernel_workspace_size<NB>(n, batch_count, output_type);
+}
+
+template <rocblas_int NB, typename Ti, typename To>
+rocblas_status rocblas_asum_batched_template(rocblas_handle  handle,
```

Comment:
I think it's probably worth it to combine the templates as well tbh, makes it a bit cleaner I think since they're identical.

Reply:
Now not sure as I think we might want to doxygen document the _template and keeping one per .cpp seems clearer to me than a single catch all.   If we make 3 reusable c++ template classes I would expect 3 .hpp files.   Can discuss.

Reply:
> 
> 
> Now not sure as I think we might want to doxygen document the _template and keeping one per .cpp seems clearer to me than a single catch all. If we make 3 reusable c++ template classes I would expect 3 .hpp files. Can discuss.

We do not need to document internal interfaces such as `*_impl` or `*_template`. Only public C interfaces need documentation. So only the stuff in `library/include/*` needs documenting, not the stuff in `library/src/*`. For the internal stuff, inline C/C++ comments are sufficient.

Reply:
> If we make 3 reusable c++ template classes I would expect 3 .hpp files.

There is no need for each template to be in its own `.hpp` file.

Templates which aren't used, don't affect program output, so if three templates are in the same header file, and only one of them is used, only the one which is used will appear in the output.

If the three template functions are somewhat related, it may be better to keep them in the same file for maintainability, rather than splitting them into separate headers, especially if the templates partially share some common code, which you don't want spread across several files.

As I said before, internal implementations do not need Doxygen documentation, just C/C++ comments.

So you can put 3 template functions in the same header, document them with only C/C++ comments, and then `#include` the header multiple times, only using a subset of it each time. Only the templates which are used (instantiated) will appear in the output.